File: src/Chocolate.cpp
#include "Chocolate.hpp"

/**
 * @brief Construct a new Chocolate:: Chocolate object
 * 
 */
Chocolate::Chocolate(){}

/**
 * @brief resets the emulator
 * 
 */
void Chocolate::reset(){
	memset(memory, 0, sizeof(memory));
	memset(V, 0, sizeof(V));
	memset(stack, 0, sizeof(stack));
	memset(keymap, 0, sizeof(keymap));

    sp = 0;
    I = 200;
    pc = 0x200;

    delay = 0;
    sound = 0;

    buzzer = false;
    draw = false;

    memset(graphics,0,2048);

    for(int i = 0; i < 80; i++){
        memory[i] = fontset[i];
    }
}

/**
 * @brief loads a ROM into memory
 * 
 * @param filename 
 * @return true file loaded successfully
 * @return false file loading failed
 */
bool Chocolate::load(const char* filename){

    reset();

    FILE* file = fopen(filename, "r");
    fread(memory + 0x200,1,sizeof(memory) - 0x200, file);
    fclose(file);

    //TODO: Come back and add validation
    return true;

}

/**
 * @brief ticks the cpu once
 * 
 */
void Chocolate::tick(){
    //Get Opcode
    op = memory[pc] << 8 | memory[pc +1 ];
  
    //Get operands
    NNN = (op & 0x0FFF);
    NN = (op & 0x00FF);
    N = (op & 0x000F);
    X = (op & 0x0F00) >> 8;
    Y = (op & 0x00F0) >> 4;

    //Reset Program Counter Delta
    pcDelta = 1;

    switch(op & 0xF000){
        case 0x0000:{
            switch(op & 0x000F){
                case 0x0000:{_00E0();}break;
                case 0x000E:{_00EE();}break;
            }
        }break;
        case 0x1000:{_1NNN();}break;
        case 0x2000:{_2NNN();}break;
        case 0x3000:{_3XNN();}break;
        case 0x4000:{_4XNN();}break;
        case 0x5000:{_5XY0();}break;
        case 0x6000:{_6XNN();}break;
        case 0x7000:{_7XNN();}break;
        case 0x8000:{
            switch(op & 0x000F){
                case 0x0000:{_8XY0();}break;
                case 0x0001:{_8XY1();}break;
                case 0x0002:{_8XY2();}break;
                case 0x0003:{_8XY3();}break;
                case 0x0004:{_8XY4();}break;
                case 0x0005:{_8XY5();}break;
                case 0x0006:{_8XY6();}break;
                case 0x0007:{_8XY7();}break;
                case 0x000E:{_8XYE();}break;
            }
        }break;
        case 0x9000:{_9XY0();}break;
        case 0xA000:{_ANNN();}break;
        case 0xB000:{_BNNN();}break;
        case 0xC000:{_CXNN();}break;
        case 0xD000:{_DXYN();}break;
        case 0xE000:{
            switch(op & 0x000F){
                case 0x000E:{_EX9E();}break;
                case 0x0001:{_EXA1();}break;
            }
        }break;
        case 0xF000:{
            switch(op & 0x0FF){
                case 0x0007:{_FX07();}break;
                case 0x000A:{_FX0A();}break;
                case 0x0015:{_FX15();}break;
                case 0x0018:{_FX18();}break;
                case 0x0029:{_FX29();}break;
                case 0x001E:{_FX1E();}break;
                case 0x0033:{_FX33();}break;
                case 0x0055:{_FX55();}break;
                case 0x0065:{_FX65();}break;
            }
        }break;
        default:{printf("unknown opcode");}
    }

    if(delay > 0)
        delay--;
    if(sound > 0)
        sound--;


    printf("%04X ",op);

    printf("NNN:%03X NN:%02X N:%1X X:%1X Y:%1X   I:%d (%03X)\n",NNN,NN,N,X,Y,I,I);

    for(int i = 0; i < 16; i++){
        printf("%d:%d ",i,V[i]);
    }
    printf("\n");

    

    pc += 2 * pcDelta;
}


/**
 * @brief Clears the screen
 * 
 */
void Chocolate::_00E0(){
    memset(graphics,0,2048);
    draw = true;
}

/**
 * @brief Return from a subroutine
 * 
 */
void Chocolate::_00EE(){
    pc = pop();             
    pcDelta = 2;               
}

/**
 * @brief Jumps to NNN
 * 
 */
void Chocolate::_1NNN(){
    pc = NNN;
    pcDelta = 0;
}

/**
 * @brief Calls subroutine at NNN
 * 
 */
void Chocolate::_2NNN(){
    push(pc);
    pc = NNN;
}

/**
 * @brief Skips the next instruction if V[X] = NN
 * 
 */

/*
void Chocolate::_3XNN(){
    pcDelta = (V[X] == NN) ? 2:1;
}*/

void Chocolate::_3XNN(){
    pcDelta = (V[X] == NN) ? 2:1;
}


/**
 * @brief Skips the next instructio if V[X] != NN
 * 
 */
void Chocolate::_4XNN(){
    pcDelta = (V[X] != NN) ? 2:1;
}


/**
 * @brief Skips the next instruction if V[X] == V[Y]
 * 
 */
void Chocolate::_5XY0(){
    pcDelta = (V[X] == V[Y]) ? 2:1;
}

/**
 * @brief Sets V[X] to NN
 * 
 */
void Chocolate::_6XNN(){
    V[X] = NN;
}

/**
 * @brief Adds NN to V[X] (Ignores Carry Flag)
 * 
 */
void Chocolate::_7XNN(){
    V[X] += NN;
}

/**
 * @brief Sets V[X] to V[Y]
 * 
 */
void Chocolate::_8XY0(){
    V[X] = V[Y];
}

/**
 * @brief Sets V[X] to V[X] | V[Y]
 * 
 */
void Chocolate::_8XY1(){
    V[X] = V[X] | V[Y];
}

/**
 * @brief Sets V[X] to V[X] & V[Y] 
 * 
 */
void Chocolate::_8XY2(){
    V[X] = V[X] & V[Y];    
}

/**
 * @brief Set V[X] to V[X] ^ V[Y]
 * 
 */
void Chocolate::_8XY3(){
    V[X] = V[X] ^ V[Y];
}

/**
 * @brief Adds V[Y] to V[X]. V[F] is set to 1 when there is a carry and 0 when there isn't
 * 
 */
void Chocolate::_8XY4(){
    V[0xF] = ((V[X] + V[Y]) > 0xFFFF) ? 1:0;
    V[X] += V[Y];
}

/**
 * @brief V[Y] is subtraced from V[X]. V[F] is set to 0 when there is a borrow and 1 where there isnt
 * 
 */
void Chocolate::_8XY5(){
    (V[Y] > V[X]) ? V[0xF]=0:V[0xF]=1;
    V[X] -= V[Y];
}

/**
 * @brief Stores the least signifigant bit of V[X] in V[F] and then shifts V[X] to the right by 1
 * 
 */
void Chocolate::_8XY6(){
    V[0xF] = V[X] & 0x0001; //Get least signifigant bit
    V[X] >>= 1;                    
}

/**
 * @brief Sets V[X] to V[Y] - V[X]. V[F] is set to 0 when there is a borrow, 1 when there isn't
 * 
 */
void Chocolate::_8XY7(){
    (V[Y] < V[X]) ? V[0xF]=0:V[0xF]=1;
    V[X] = V[Y] - V[X];
}

/**
 * @brief Stores the most signifigant bit of V[X] in V[F] then shifts V[F] to the left by 1
 * 
 */
void Chocolate::_8XYE(){
    V[0xF] = V[X] & 0x80; //Get most signifigant bit
    V[X] <<= 1;
}

/**
 * @brief Skips the next instruction if V[X] != V[Y]
 * 
 */
void Chocolate::_9XY0(){
    (V[X] != V[Y]) ? pcDelta =2:pcDelta=1;
}

/**
 * @brief Sets I to the address NNN
 * 
 */
void Chocolate::_ANNN(){
    I = NNN;
}

/**
 * @brief Jups to the address NNN plus V[0]
 * 
 */
void Chocolate::_BNNN(){
    I = NNN + V[0];
}

/**
 * @brief Sets V[X] to the result of a bitwise and operation on a random number (Typically 0 to 255) and NN
 * 
 */
void Chocolate::_CXNN(){
    V[X] = 1;
    //V[X] = (rand() % (0xFF +1)) & NN;
}

/**
 * @brief Draws a sprite at (V[X],V[Y]) that has a width of 8 pixels and a height of N pixels
 * Each row of 8 blices is read as a bit-coded from memory location I; I value doesn't change after 
 * execution. As described above V[F] is set to 1 if any pixels are flipped from the set to unset when the
 * sprite is drawn, and to 0 if that doesn't happen
 */
void Chocolate::_DXYN(){

    uint16_t pixel;
    
    V[0xF] = 0;

    for(int i = 0; i < N; i++){
        pixel = memory[I+i];
        for(int j = 0; j < 8; j++){
            if((pixel & (0x80 >> j)) != 0){
                if(graphics[(X + j + ((Y+ i) * 64)) ] == 1){
                    V[0xF] = 1;
                }
                graphics[X + j + ((Y+i) * 64)] ^= 1;
            }
        }
    }
    draw = true;
}


/**
 * @brief Skips the next instruction if the key stored in V[X] is pressed
 * 
 */
void Chocolate::_EX9E(){
    pcDelta = (keymap[V[X]] != 0) ? 2:1;    
}

/**
 * @brief Skips the next instruction if the key stored in V[X] isn't pressed
 * 
 */
void Chocolate::_EXA1(){
    pcDelta = (keymap[V[X]] != 0) ? 2:1;
}

/**
 * @brief Sets V[X] to the value of the delay timer
 * 
 */
void Chocolate::_FX07(){
    V[X] = delay;
}

/**
 * @brief A key press is awaited, and then stored in V[X]
 * 
 */
void Chocolate::_FX0A(){
    bool keypress = false;

    for(int i = 0; i < 16; i++){
        if(keymap[i] != 0){
            V[X] = i;
            keypress = true;
        }
    }
    if(!keypress){
        pcDelta = 0;
    }
}

/**
 * @brief Sets the delay timer to V[X]
 * 
 */
void Chocolate::_FX15(){
    delay = V[X];    
}

/**
 * @brief Sets the sound timer to V[X]
 * 
 */
void Chocolate::_FX18(){
    sound = V[X];
}

/**
 * @brief Adds V[X] to I
 * 
 */
void Chocolate::_FX1E(){
    I += V[X];
}

/**
 * @brief Sets I to the location of the sprice to the character in V[X] (0-F)
 * 
 */
void Chocolate::_FX29(){
    I = V[X] * 0x5;
}

/**
 * @brief Stores the BCD respresentation of V[X]. 100's digit at I, 10's at I+1, I's at I+2
 */
void Chocolate::_FX33(){
        memory[I] = V[X] /100;
        memory[I+1] = (V[X] /10)%10;
        memory[I+2] = V[X] % 10;
}

/**
 * @brief Stores V[0] to V[X] in memory starting at address I. The offset from I is increased
 * by 1 for each value written but I itself is left unmodified
 */
void Chocolate::_FX55(){  
    for(int i = 0; i <= V[X]; i++){
        memory[I + i] = V[i];
    }
}

/**
 * @brief Fills V[0] to V[X] with falues from memory starting at address I. The offset from I
 * is increased by 1 for each value written, but I itself is left unmodified 
 * 
 */
void Chocolate::_FX65(){
    for(int i = 0; i <= V[X]; i++){
        V[i] = memory[I+i];
    }
}


/**
 * @brief pushes a value onto the stack
 * 
 * @param val 
 */
void Chocolate::push(uint16_t val){
    stack[sp++] = val;
}

/**
 * @brief pops the value on the top of the stack
 * 
 * @return uint16_t 
 */
uint16_t Chocolate::pop(){
    return stack[sp--];
}

File: src/chocolate.rs
const FONTSET: [u8; 80] = 
[
    0xF0, 0x90, 0x90, 0x90, 0xF0, //0
    0x20, 0x60, 0x20, 0x20, 0x70, //1
    0xF0, 0x10, 0xF0, 0x80, 0xF0, //2
    0xF0, 0x10, 0xF0, 0x10, 0xF0, //3
    0x90, 0x90, 0xF0, 0x10, 0x10, //4
    0xF0, 0x80, 0xF0, 0x10, 0xF0, //5
    0xF0, 0x80, 0xF0, 0x90, 0xF0, //6
    0xF0, 0x10, 0x20, 0x40, 0x40, //7
    0xF0, 0x90, 0xF0, 0x90, 0xF0, //8
    0xF0, 0x90, 0xF0, 0x10, 0xF0, //9
    0xF0, 0x90, 0xF0, 0x90, 0x90, //A
    0xE0, 0x90, 0xE0, 0x90, 0xE0, //B
    0xF0, 0x80, 0x80, 0x80, 0xF0, //C
    0xE0, 0x90, 0x90, 0x90, 0xE0, //D
    0xF0, 0x80, 0xF0, 0x80, 0xF0, //E
    0xF0, 0x80, 0xF0, 0x80, 0x80  //F
];


pub struct Chocolate {
    memory: [u8; 0xFFF], // All Available Memory
    v: [u8; 16],  // Registers
    stack: [u16; 16], // Stack
    sp: u8, // Stack Pointer
    i: u16, // 16 bit register
    pc: u16, // Program counter
    delay: u8, // Delay Timer
    sound: u8, //Buzzer Timer
    buzzer: bool, //If buzzer is active
    draw: bool, //A boolean to re-draw the screen
    keymap: [u8; 16], // Stores what keys are being pressed
    graphics: [u8; 2048], //Screen Data
    pc_delta: i16 //Internal Program Counter increase delta
}


impl Chocolate {

    pub fn new() -> Chocolate {
        let mut chocolate = Chocolate{
            memory: [0; 0xFFF],
            v: [0; 16],
            stack: [0; 16],
            sp: 0,
            i: 0,
            pc: 0,
            delay: 0,
            sound: 0,
            buzzer: false,
            draw: false,
            keymap: [0; 16],
            graphics: [0; 2048],
            pc_delta: 1
        };
        for i in 0..80 {
            chocolate.memory[i] = FONTSET[i];
        }
        return chocolate;
    }

    /// Loads the ROM into memory
    fn load(&mut self, rom: &[i8]){
        for(i, value) in rom.iter().enumerate() {
            self.memory[0x200 + i] = *value as u8; 
        }
    }


    // Ticks the machine one clock cycle
    pub fn tick(&mut self, debug: bool) {
        //Get opcode

        let op = (self.memory[self.pc as usize] as u16) << 8 | self.memory[(self.pc +1) as usize] as u16;
        
        // Get operands
        let nnn = op & 0x0FFF;
        let kk = (op & 0x00FF) as u8;
        let n = (op & 0x000F) as u8;
        let x = ((op & 0x0F00) >> 8) as u8;
        let y = ((op & 0x00F0) >> 4) as u8;

        if debug {
            println!("{:#04X} : {:#04X}", self.pc, op);
            println!("NNN:{:#03X} KK:{:#02X} N:{:#X} X:{:#X} Y:{:#X}   I:{} ({:#03X})", nnn, kk, n, x, y, self.i, self.i);

        }

        //Reset the program counter delta
        self.pc_delta = 1;
        match op & 0x00FF {
            0x00E0 => self.cls(),
            0x00EE => self.ret(),
            0x0000..=0x01FF => println!("SYS instruction ignored"),
            0x0200..=0x0FFF => println!("Unimplemented Opcode"),
            0x1000..=0x1FFF => self.jp(nnn),
            0x2000..=0x2FFF => self.call(nnn),
            0x3000..=0x3FFF => self.se(x, kk),
            0x4000..=0x4FFF => self.sne(x, kk),
            0x5000..=0x5FFF => self.se_reg(x, y),
            0x6000..=0x6FFF => self.ld(x, kk),
            0x7000..=0x7FFF => self.add(x, kk),
            0x8000..=0x8FFF => {
                match op & 0x000F {
                    0x0000 => self.ld_reg_reg(x, y),
                    0x0001 => self.or(x, y),
                    0x0002 => self.and(x, y),
                    0x0003 => self.xor(x, y),
                    0x0004 => self.add_carry(x, y),
                    0x0005 => self.sub(x, y),
                    0x0006 => self.shr(x),
                    0x0007 => self.subn(x, y),
                    0x000E => self.shl(x),
                    _ => println!("Unknown Opcode"),
                }
            },
            0x9000..=0x9FFF => self.sne_register(x, y),
            0xA000..=0xAFFF => self.ld_i(nnn),
            0xB000..=0xBFFF => self.jmp_memory(nnn),
            0xC000..=0xCFFF => self.rnd(x, kk),
            0xD000..=0xDFFF => self.drw(x,y,n),
            0xE09E => self.skp(x),
            0xE0A1 => self.sknp(x),
            0xF007 => self.ld_reg_delay(x),
            0xF00A => self.ld_key(x),
            0xF015 => self.ld_delay_reg(x),
            0xF018 => self.ld_sound(x),
            0xF01E => self.add_i(x),
            0xF029 => self.ld_font(x),
            0xF033 => self.ld_bcd(x),
            0xF055 => self.ld_mem(x),
            0xF065 => self.ld_reg_mem(x),
            _ => println!("Unknown Opcode"),
        }

        if self.delay > 0 {
            self.delay -= 1;
        }

        if self.sound > 0 {
            self.sound -= 1; 
        }

    }

    /// # ``00E0`` - CLS
    /// Clear the display. 
    fn cls(&mut self) {
        self.graphics.fill(0);
        self.draw = true; 
    }

    /// # ``00EE`` - RET
    /// Return from a subroutine.
    /// The interpreter sets the program counter to the address at the top of the stack, then subtracts 1 from the stack pointer.
    fn ret(&mut self) {
        self.sp -= 1; 
        self.pc = self.stack[self.sp as usize];
    }

    /// # ``1nnn`` - JP addr
    /// The interpreter sets the program counter to nnn.
    /// ## Arguments
    /// - ``nnn`` Jump to location nnn.
    fn jp(&mut self, nnn: u16){
        self.pc = nnn;
        self.pc_delta = 0; 
    }

    /// # ``2nnn`` - CALL addr
    /// The interpreter increments the stack pointer, then puts the current PC on the top of the stack. The PC is then set to nnn.
    /// ## Arguments
    /// - ``nnn`` Call subroutine at nnn
    fn call(&mut self, nnn: u16){
        self.stack[self.sp as usize] = self.pc;
        self.sp += 1;
        self.pc = nnn;
        self.pc_delta = 0; 
    }

    /// # ``3xkk`` - SE Vx, byte
    /// The interpreter compares register Vx to nn, and if they are equal, increments the program counter by 2.
    /// ## Arguments
    /// - ``x`` Register to compare
    /// - ``nn`` Value to compare
    fn se(&mut self, x: u8, kk: u8){
       if self.v[x as usize] == kk {self.pc_delta = 2}; 
    }

    /// # ``4xkk`` - SNE Vx, byte
    /// The interpreter compares register Vx to kk, and if they are not equal, increments the program counter by 2.
    /// ## Arguments
    /// - ``x`` Register to compare
    /// - ``nn`` Value to compare
    fn sne(&mut self, x: u8, kk:u8){
        if self.v[x as usize] != kk {self.pc_delta = 2}; 
    }

    /// # ``5xy0`` - SE Vx, Vy
    /// The interpreter compares register Vx to register Vy, and if they are equal, increments the program counter by 2.
    /// ## Arguments
    /// - ``x`` Register to compare
    /// - ``y`` Register to compare
    fn se_reg(&mut self, x: u8, y: u8){
        if self.v[x as usize] == self.v[y as usize] {self.pc_delta = 2}
    }

    /// # ``6xkk`` - LD Vx, byte
    /// The interpreter puts the value kk into register Vx
    /// ## Arguments
    /// - ``x`` Register to fill
    /// - ``kk`` Value
    fn ld(&mut self, x: u8, kk: u8){
        self.v[x as usize] = kk;     
    }

    /// # ``7xkk`` - ADD Vx, byte
    /// Set Vx = Vx + kk.
    /// Adds the value kk to the value of register Vx, then stores the result in Vx. 
    /// ## Arguments
    /// - ``x`` register to add
    /// - ``kk`` value
    fn add(&mut self, x: u8, kk: u8){
        self.v[x as usize] += kk
    }

    /// # ``8xy0`` - LD Vx, Vy
    /// Set Vx = Vy.
    /// Stores the value of register Vy in register Vx.
    /// ## Arguments
    /// - ``x`` source register
    /// - ``y`` destination register 
    fn ld_reg_reg(&mut self, x: u8, y: u8){
        self.v[x as usize] = self.v[y as usize] 
    }

    /// # ``8xy1`` - OR Vx, Vy
    /// Set Vx = Vx OR Vy.
    /// Performs a bitwise OR on the values of Vx and Vy, then stores the result in Vx. 
    /// A bitwise OR compares the corrseponding bits from two values, 
    /// and if either bit is 1, then the same bit in the result is also 1. Otherwise, it is 0.
    /// ## Arguments
    /// - `x` Destination register
    /// - `y` Check register
    fn or(&mut self, x: u8, y: u8){

    }

    /// # ``8xy2`` - AND Vx, Vy
    /// Set Vx = Vx AND Vy.
    /// Performs a bitwise AND on the values of Vx and Vy, then stores the result in Vx. A bitwise AND compares the corrseponding bits from two values, and if both bits are 1, then the same bit in the result is also 1. Otherwise, it is 0.
    /// ## Arguments
    /// - `x` Destination register
    /// - `y` Check register
    fn and(&mut self, x: u8, y: u8){

    }

    /// # `8xy3` - XOR Vx, Vy
    /// Set Vx = Vx XOR Vy.
    /// Performs a bitwise exclusive OR on the values of Vx and Vy, then stores the result in Vx. An exclusive OR compares the corrseponding bits from two values, and if the bits are not both the same, then the corresponding bit in the result is set to 1. Otherwise, it is 0. 
    /// ## Arguments
    /// - `x` Destination register
    /// - `y` Check register
    fn xor(&mut self, x: u8, y: u8){

    }

    /// # `8xy4` - ADD Vx, Vy
    /// Set Vx = Vx + Vy, set VF = carry.
    /// The values of Vx and Vy are added together. If the result is greater than 8 bits (i.e., > 255,) VF is set to 1, otherwise 0. Only the lowest 8 bits of the result are kept, and stored in Vx.
    /// ## Argument
    /// - `x` Destination Register
    /// - `y` Operand Register
    fn add_carry(&mut self, x: u8, y: u8){

    }


    /// # `8xy5` - SUB Vx, Vy
    /// Set Vx = Vx - Vy, set VF = NOT borrow.
    /// If Vx > Vy, then VF is set to 1, otherwise 0. Then Vy is subtracted from Vx, and the results stored in Vx.
    /// ## Arguments
    /// - `x` Destination Register 
    /// - `y` Operand Register
    fn sub(&mut self, x: u8, y: u8){

    }

    /// # `8xy6` - SHR Vx {, Vy}
    /// Set Vx = Vx SHR 1.
    /// If the least-significant bit of Vx is 1, then VF is set to 1, otherwise 0. Then Vx is divided by 2.
    /// ## Arguments
    /// - `x` Register
    fn shr(&mut self, x: u8){

    }

    /// # `8xy7` - SUBN Vx, Vy
    /// Set Vx = Vy - Vx, set VF = NOT borrow.
    /// If Vy > Vx, then VF is set to 1, otherwise 0. Then Vx is subtracted from Vy, and the results stored in Vx.
    /// ## Arguments
    /// - `x` Destination Register
    /// - `y` Operand Register
    fn subn(&mut self, x: u8, y: u8){

    }

    /// # `8xyE` - SHL Vx {, Vy}
    /// Set Vx = Vx SHL 1.
    /// If the most-significant bit of Vx is 1, then VF is set to 1, otherwise to 0. Then Vx is multiplied by 2.
    /// ## Arguments
    /// - `x` Register
    fn shl(&mut self, x: u8){

    }

    /// # `9xy0` - SNE Vx, Vy
    /// Skip next instruction if Vx != Vy.
    /// The values of Vx and Vy are compared, and if they are not equal, the program counter is increased by 2.
    /// ## Arguments
    /// - `x` First Register
    /// - `y` Second Register
    fn sne_register(&mut self, x: u8, y: u8){

    }

    /// # `Annn` - LD I, addr
    /// Set I = nnn.
    /// The value of register I is set to nnn.
    /// ## Arguments
    /// - `nnn` Address
    fn ld_i(&mut self,  nnn: u16){

    }

    /// # `Bnnn` - JP V0, addr
    /// Jump to location nnn + V0.
    /// The program counter is set to nnn plus the value of V0.
    /// ## Arguments
    /// - `nnn` Address
    fn jmp_memory(&mut self, nnn: u16){

    }

    /// # Cxkk - RND Vx, byte
    /// Set Vx = random byte AND kk.
    /// The interpreter generates a random number from 0 to 255, which is then ANDed with the value kk. The results are stored in Vx. See instruction 8xy2 for more information on AND.
    /// ## Arguments
    /// - `x` Register
    /// - `kk` Value 
    fn rnd(&mut self, x: u8, kk: u8){

    }

    /// # `Dxyn` - DRW Vx, Vy, nibble
    /// Display n-byte sprite starting at memory location I at (Vx, Vy), set VF = collision.
    /// The interpreter reads n bytes from memory, starting at the address stored in I. These bytes are then displayed as sprites on screen at coordinates (Vx, Vy). Sprites are XORed onto the existing screen. If this causes any pixels to be erased, VF is set to 1, otherwise it is set to 0. If the sprite is positioned so part of it is outside the coordinates of the display, it wraps around to the opposite side of the screen. See instruction 8xy3 for more information on XOR, and section 2.4, Display, for more information on the Chip-8 screen and sprites.
    /// ## Arguments
    /// - `x` Register for x coordinate
    /// - `y` Register for y coordinate
    /// - `n` bytes
    fn drw(&mut self, x: u8, y: u8, n: u8){

    }

    /// # `Ex9E` - SKP Vx
    /// Skip next instruction if key with the value of Vx is pressed.
    /// Checks the keyboard, and if the key corresponding to the value of Vx is currently in the down position, PC is increased by 2.
    /// ## Arguments
    /// - `x` key value
    fn skp(&mut self, x: u8){

    }

    /// # `ExA1` - SKNP Vx  
    /// Skip next instruction if key with the value of Vx is not pressed.
    /// Checks the keyboard, and if the key corresponding to the value of Vx is currently in the up position, PC is increased by 2.
    /// ## Arguments
    /// - `x` key value
    fn sknp(&mut self, x: u8){

    }

    /// # `Fx07` - LD Vx, DT
    /// Set Vx = delay timer value.
    /// The value of DT is placed into Vx.
    /// ## Arguments
    /// - `x` Register
    fn ld_reg_delay(&mut self, x: u8){

    }

    /// # `Fx0A` - LD Vx, K
    /// Wait for a key press, store the value of the key in Vx.
    /// All execution stops until a key is pressed, then the value of that key is stored in Vx.
    /// ## Arguments
    /// - `x` Register
    fn ld_key(&mut self, x:u8){

    } 

    /// # `Fx15` - LD DT, Vx
    /// Set delay timer = Vx.
    /// DT is set equal to the value of Vx.
    /// ## Arguments
    /// - `x` Register
    fn ld_delay_reg(&mut self, x:u8){

    }

    /// # `Fx18` - LD ST, Vx
    /// Set sound timer = Vx.
    /// ST is set equal to the value of Vx.
    /// ## Arguments
    /// - `x` Register
    fn ld_sound(&mut self, x: u8){

    }

    /// # `Fx1E` - ADD I, Vx
    /// Set I = I + Vx.
    /// The values of I and Vx are added, and the results are stored in I.
    /// ## Arguments
    /// - `x` Register
    fn add_i(&mut self, x: u8){

    }

    /// # `Fx29` - LD F, Vx
    /// Set I = location of sprite for digit Vx.
    /// The value of I is set to the location for the hexadecimal sprite corresponding to the value of Vx. See section 2.4, Display, for more information on the Chip-8 hexadecimal font.
    /// ## Arguments
    /// - `x` Register
    fn ld_font(&mut self, x: u8){

    }

    /// # `Fx33` - LD B, Vx
    /// Store BCD representation of Vx in memory locations I, I+1, and I+2.
    /// The interpreter takes the decimal value of Vx, and places the hundreds digit in memory at location in I, the tens digit at location I+1, and the ones digit at location I+2.
    /// ## Arguments
    /// - `x` Register
    fn ld_bcd(&mut self, x: u8){

    }

    /// # `Fx55` - LD [I], Vx
    /// Store registers V0 through Vx in memory starting at location I.
    /// The interpreter copies the values of registers V0 through Vx into memory, starting at the address in I.
    /// ## Arguments
    /// - `x` Register
    fn ld_mem(&mut self, x: u8){

    }

    /// # `Fx65` - LD Vx, [I]
    /// Read registers V0 through Vx from memory starting at location I.
    /// The interpreter reads values from memory starting at location I into registers V0 through Vx.
    /// ## Arguments
    /// - `x` Register
    fn ld_reg_mem(&mut self, x: u8){

    }




}
